File: src/shapes/trapezoid.js
// Trapezoid shape geometry calculations
export const trapezoidGeometry = {
  // Generate SVG path for trapezoid
  generatePath: (topWidth, bottomWidth, height) => {
    // Calculate the horizontal offset for centering top width
    const offset = (bottomWidth - topWidth) / 2;
    
    // Trapezoid vertices (clockwise from top-left)
    const vertices = [
      { x: offset, y: 0 },                    // Top-left
      { x: offset + topWidth, y: 0 },        // Top-right
      { x: bottomWidth, y: height },          // Bottom-right
      { x: 0, y: height }                     // Bottom-left
    ];
    
    // Create path string
    const pathString = vertices
      .map((vertex, index) => {
        const command = index === 0 ? 'M' : 'L';
        return `${command} ${vertex.x} ${vertex.y}`;
      })
      .join(' ') + ' Z';
    
    return pathString;
  },

  // Calculate bounding box
  getBoundingBox: (topWidth, bottomWidth, height) => {
    const maxWidth = Math.max(topWidth, bottomWidth);
    const minX = bottomWidth > topWidth ? 0 : (topWidth - bottomWidth) / 2;
    
    return {
      minX: minX,
      minY: 0,
      maxX: minX + maxWidth,
      maxY: height,
      width: maxWidth,
      height: height
    };
  },

  // Calculate area
  getArea: (topWidth, bottomWidth, height) => {
    return ((topWidth + bottomWidth) / 2) * height;
  },

  // Calculate perimeter
  getPerimeter: (topWidth, bottomWidth, height) => {
    const offset = (bottomWidth - topWidth) / 2;
    const sideLength = Math.sqrt(Math.pow(offset, 2) + Math.pow(height, 2));
    
    return topWidth + bottomWidth + (2 * sideLength);
  },

  // Get center point
  getCenter: (topWidth, bottomWidth, height) => {
    const offset = (bottomWidth - topWidth) / 2;
    const centerX = bottomWidth / 2;
    const centerY = height / 2;
    
    return {
      x: centerX,
      y: centerY
    };
  },

  // Get vertices
  getVertices: (topWidth, bottomWidth, height) => {
    const offset = (bottomWidth - topWidth) / 2;
    
    return [
      { x: offset, y: 0 },                    // Top-left
      { x: offset + topWidth, y: 0 },        // Top-right
      { x: bottomWidth, y: height },          // Bottom-right
      { x: 0, y: height }                     // Bottom-left
    ];
  },

  // Get dimension points for labels
  getDimensionPoints: (topWidth, bottomWidth, height) => {
    const offset = (bottomWidth - topWidth) / 2;
    
    return {
      topWidth: {
        start: { x: offset, y: -20 },
        end: { x: offset + topWidth, y: -20 },
        label: { x: offset + topWidth / 2, y: -35 }
      },
      bottomWidth: {
        start: { x: 0, y: height + 20 },
        end: { x: bottomWidth, y: height + 20 },
        label: { x: bottomWidth / 2, y: height + 35 }
      },
      height: {
        start: { x: bottomWidth + 20, y: 0 },
        end: { x: bottomWidth + 20, y: height },
        label: { x: bottomWidth + 35, y: height / 2 }
      },
      leftSide: {
        start: { x: -20, y: 0 },
        end: { x: -20, y: height },
        label: { x: -35, y: height / 2 }
      },
      rightSide: {
        start: { x: bottomWidth + 20, y: 0 },
        end: { x: bottomWidth + 20, y: height },
        label: { x: bottomWidth + 35, y: height / 2 }
      }
    };
  },

  // Calculate side lengths
  getSideLengths: (topWidth, bottomWidth, height) => {
    const offset = Math.abs(bottomWidth - topWidth) / 2;
    const sideLength = Math.sqrt(Math.pow(offset, 2) + Math.pow(height, 2));
    
    return {
      top: topWidth,
      bottom: bottomWidth,
      left: sideLength,
      right: sideLength
    };
  },

  // Calculate angles
  getAngles: (topWidth, bottomWidth, height) => {
    const offset = Math.abs(bottomWidth - topWidth) / 2;
    
    // Calculate base angles
    const leftAngleRad = Math.atan2(height, offset);
    const rightAngleRad = Math.atan2(height, offset);
    
    return {
      topLeft: (leftAngleRad * 180) / Math.PI,
      topRight: (rightAngleRad * 180) / Math.PI,
      bottomLeft: 180 - ((leftAngleRad * 180) / Math.PI),
      bottomRight: 180 - ((rightAngleRad * 180) / Math.PI)
    };
  },

  // Validate parameters
  validateParameters: (topWidth, bottomWidth, height) => {
    const errors = [];
    
    if (topWidth <= 0) errors.push('Top width must be greater than 0');
    if (bottomWidth <= 0) errors.push('Bottom width must be greater than 0');
    if (height <= 0) errors.push('Height must be greater than 0');
    
    if (topWidth > 10000) errors.push('Top width must be less than 10000mm');
    if (bottomWidth > 10000) errors.push('Bottom width must be less than 10000mm');
    if (height > 10000) errors.push('Height must be less than 10000mm');
    
    // Check if trapezoid is geometrically valid
    const maxWidth = Math.max(topWidth, bottomWidth);
    const minWidth = Math.min(topWidth, bottomWidth);
    
    if (minWidth <= 0) errors.push('Both widths must be positive');
    
    return {
      isValid: errors.length === 0,
      errors
    };
  },

  // Generate different trapezoid types
  generateIsoscelesTrapezoid: (topWidth, bottomWidth, height) => {
    // This is the standard trapezoid (already implemented)
    return {
      path: this.generatePath(topWidth, bottomWidth, height),
      type: 'isosceles'
    };
  },

  generateRightTrapezoid: (topWidth, bottomWidth, height, isLeftRight = false) => {
    // Right trapezoid with one vertical side
    if (isLeftRight) {
      // Left side is vertical
      return {
        path: `M 0 0 L ${topWidth} 0 L ${bottomWidth} ${height} L 0 ${height} Z`,
        type: 'right-left'
      };
    } else {
      // Right side is vertical
      return {
        path: `M ${bottomWidth - topWidth} 0 L ${bottomWidth} 0 L ${bottomWidth} ${height} L ${bottomWidth - topWidth} ${height} Z`,
        type: 'right-right'
      };
    }
  },

  // Calculate centroid (geometric center)
  getCentroid: (topWidth, bottomWidth, height) => {
    // For an isosceles trapezoid
    const offset = (bottomWidth - topWidth) / 2;
    const centroidX = bottomWidth / 2;
    const centroidY = height * (2 * topWidth + bottomWidth) / (3 * (topWidth + bottomWidth));
    
    return {
      x: centroidX,
      y: centroidY
    };
  }
};

export default trapezoidGeometry;

File: src-tauri/src/shapes/trapezoid.rs
use crate::shapes::{BoundingBox, Dimension, DimensionOrientation, Point, ShapeGeometry, Transform, ValidationError, ShapeParameters};

pub struct TrapezoidGeometry;

impl ShapeGeometry for TrapezoidGeometry {
    fn generate_path(&self, params: &ShapeParameters) -> String {
        let top_width = params.top_width.unwrap_or(80.0);
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        // Calculate horizontal offsets to center the top width
        let bottom_offset = (bottom_width - top_width) / 2.0;
        
        // Trapezoid vertices (clockwise from top-left)
        let vertices = vec![
            (bottom_offset, 0.0),                    // Top-left
            (bottom_offset + top_width, 0.0),         // Top-right
            (bottom_width, height),                   // Bottom-right
            (0.0, height),                            // Bottom-left
        ];
        
        let mut path_parts = Vec::new();
        for (i, (x, y)) in vertices.iter().enumerate() {
            if i == 0 {
                path_parts.push(format!("M {} {}", x, y));
            } else {
                path_parts.push(format!("L {} {}", x, y));
            }
        }
        path_parts.push("Z".to_string());
        
        path_parts.join(" ")
    }

    fn get_bounding_box(&self, params: &ShapeParameters) -> BoundingBox {
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        BoundingBox {
            min_x: 0.0,
            min_y: 0.0,
            max_x: bottom_width,
            max_y: height,
            width: bottom_width,
            height,
        }
    }

    fn get_area(&self, params: &ShapeParameters) -> f64 {
        let top_width = params.top_width.unwrap_or(80.0);
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        // Area of trapezoid: (a + b) / 2 * h
        ((top_width + bottom_width) / 2.0) * height
    }

    fn get_perimeter(&self, params: &ShapeParameters) -> f64 {
        let top_width = params.top_width.unwrap_or(80.0);
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        // Calculate the length of the slanted sides
        let bottom_offset = (bottom_width - top_width) / 2.0;
        let side_length = (bottom_offset * bottom_offset + height * height).sqrt();
        
        top_width + bottom_width + 2.0 * side_length
    }

    fn get_center(&self, params: &ShapeParameters) -> Point {
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        // Center of trapezoid (approximate)
        Point {
            x: bottom_width / 2.0,
            y: height / 2.0,
        }
    }

    fn get_vertices(&self, params: &ShapeParameters) -> Vec<Point> {
        let top_width = params.top_width.unwrap_or(80.0);
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        
        let bottom_offset = (bottom_width - top_width) / 2.0;
        
        vec![
            Point { x: bottom_offset, y: 0.0 },                    // Top-left
            Point { x: bottom_offset + top_width, y: 0.0 },         // Top-right
            Point { x: bottom_width, y: height },                   // Bottom-right
            Point { x: 0.0, y: height },                           // Bottom-left
        ]
    }

    fn validate_parameters(&self, params: &ShapeParameters) -> ValidationError {
        let mut errors = Vec::new();
        
        if let Some(top_width) = params.top_width {
            if top_width <= 0.0 {
                errors.push("Top width must be greater than 0".to_string());
            }
            if top_width > 10000.0 {
                errors.push("Top width must be less than 10000mm".to_string());
            }
        } else {
            errors.push("Top width is required".to_string());
        }
        
        if let Some(bottom_width) = params.bottom_width {
            if bottom_width <= 0.0 {
                errors.push("Bottom width must be greater than 0".to_string());
            }
            if bottom_width > 10000.0 {
                errors.push("Bottom width must be less than 10000mm".to_string());
            }
        } else {
            errors.push("Bottom width is required".to_string());
        }
        
        if let Some(height) = params.height {
            if height <= 0.0 {
                errors.push("Height must be greater than 0".to_string());
            }
            if height > 10000.0 {
                errors.push("Height must be less than 10000mm".to_string());
            }
        } else {
            errors.push("Height is required".to_string());
        }
        
        // Validate that widths are reasonable relative to each other
        if let (Some(top_width), Some(bottom_width)) = (params.top_width, params.bottom_width) {
            if top_width > bottom_width * 2.0 || bottom_width > top_width * 2.0 {
                errors.push("Top and bottom widths should be reasonably proportional".to_string());
            }
        }
        
        ValidationError {
            is_valid: errors.is_empty(),
            errors,
        }
    }

    fn get_dimensions(&self, params: &ShapeParameters, render_offset: &Point, transform: &Transform) -> Vec<Dimension> {
        let top_width = params.top_width.unwrap_or(80.0);
        let bottom_width = params.bottom_width.unwrap_or(120.0);
        let height = params.height.unwrap_or(60.0);
        let bottom_offset = (bottom_width - top_width) / 2.0;
        let offset = 20.0;
        
        // Create base dimensions (before transformation)
        let base_dimensions = vec![
            // Top width dimension (horizontal at top)
            Dimension {
                start_point: Point { x: bottom_offset + render_offset.x, y: -offset + render_offset.y },
                end_point: Point { x: bottom_offset + top_width + render_offset.x, y: -offset + render_offset.y },
                text_position: Point { x: bottom_offset + top_width / 2.0 + render_offset.x, y: -offset - 10.0 + render_offset.y },
                value: top_width,
                label: format!("{:.0}mm", top_width),
                orientation: DimensionOrientation::Horizontal,
            },
            // Bottom width dimension (horizontal at bottom)
            Dimension {
                start_point: Point { x: 0.0 + render_offset.x, y: height + offset + render_offset.y },
                end_point: Point { x: bottom_width + render_offset.x, y: height + offset + render_offset.y },
                text_position: Point { x: bottom_width / 2.0 + render_offset.x, y: height + offset + 15.0 + render_offset.y },
                value: bottom_width,
                label: format!("{:.0}mm", bottom_width),
                orientation: DimensionOrientation::Horizontal,
            },
            // Height dimension (vertical at center)
            Dimension {
                start_point: Point { x: bottom_width / 2.0 + offset + render_offset.x, y: 0.0 + render_offset.y },
                end_point: Point { x: bottom_width / 2.0 + offset + render_offset.x, y: height + render_offset.y },
                text_position: Point { x: bottom_width / 2.0 + offset + 15.0 + render_offset.x, y: height / 2.0 + render_offset.y },
                value: height,
                label: format!("{:.0}mm", height),
                orientation: DimensionOrientation::Vertical,
            },
        ];
        
        // Apply rotation transformation if needed
        if transform.rotation != 0.0 {
            let center = self.get_center(params);
            let adjusted_center = Point {
                x: center.x + render_offset.x,
                y: center.y + render_offset.y,
            };
            
            base_dimensions.into_iter().map(|mut dim| {
                dim.start_point = self.transform_point(&dim.start_point, &adjusted_center, transform);
                dim.end_point = self.transform_point(&dim.end_point, &adjusted_center, transform);
                dim.text_position = self.transform_point(&dim.text_position, &adjusted_center, transform);
                
                // Update dimension orientation based on rotation
                let rotation_degrees = transform.rotation;
                let normalized_rotation = rotation_degrees % 360.0;
                
                if (normalized_rotation >= 45.0 && normalized_rotation < 135.0) || 
                   (normalized_rotation >= 225.0 && normalized_rotation < 315.0) {
                    // Swap orientations for 90° and 270° rotations
                    match dim.orientation {
                        DimensionOrientation::Horizontal => dim.orientation = DimensionOrientation::Vertical,
                        DimensionOrientation::Vertical => dim.orientation = DimensionOrientation::Horizontal,
                        _ => {}
                    }
                }
                
                dim
            }).collect()
        } else {
            base_dimensions
        }
    }
    
    fn transform_point(&self, point: &Point, center: &Point, transform: &Transform) -> Point {
        let mut x = point.x;
        let mut y = point.y;

        // Apply rotation
        if transform.rotation != 0.0 {
            let angle_rad = (transform.rotation * std::f64::consts::PI) / 180.0;
            let cos_a = angle_rad.cos();
            let sin_a = angle_rad.sin();
            
            let rel_x = x - center.x;
            let rel_y = y - center.y;
            
            x = rel_x * cos_a - rel_y * sin_a + center.x;
            y = rel_x * sin_a + rel_y * cos_a + center.y;
        }

        Point { x, y }
    }

    fn get_rotation_center(&self, params: &ShapeParameters) -> Point {
        // For trapezoid, use the same center as the shape
        self.get_center(params)
    }
}
